### PR TITLE
Update struct StorageServiceDeployment for API version "2014-06-01"

### PR DESCRIPTION
**Pull request by dima-exe**
The SecondaryReadEnabled and the GeoReplicationEnabled element is only
available using version 2013-11-01 or higher and is replaced by the
AccountType element with version 2014-06-01 or higher.

### CODE REVIEW

File: management/storageservice/entities.go [hunk 1]
```diff
@@ -34,16 +34,15 @@ type StorageServiceProperties struct {
...
-	SecondaryReadEnabled  bool
+	XMLName            xml.Name `xml:"CreateStorageServiceInput"`
+	Xmlns              string   `xml:"xmlns,attr"`
+	ServiceName        string
+	Description        string
+	Label              string
+	AffinityGroup      string `xml:",omitempty"`
+	Location           string `xml:",omitempty"`
+	ExtendedProperties ExtendedPropertyList
+	AccountType        string
```
Comment (ahmetalpbalkan) [inline management/storageservice/entities.go]: Can you please move AccountType up so that ExtendedProperties appear last?

**Comment by ahmetalpbalkan** [comment]
Thanks for the PR @dima-exe. Does this make all accounts generated by that method GRS? Ideally we'd love to accept the struct as input not just "name & "location". Unfortunately the code is a bit hacky at the moment. Would you be kind enough to make it so?

**Comment by dima-exe** [comment]
@ahmetalpbalkan Please check the last commit.

**Comment by ahmetalpbalkan** [comment]
@dima-exe you just expanded scope of this PR. :smile: I was expecting you to just fix the storageservice part. @paulmey was also up to a similar refactor (paul can you please review as well)?

File: management/storageservice/client.go [hunk 2]
```diff
@@ -84,7 +84,7 @@ func (self StorageServiceClient) GetStorageServiceByLocation(location string) (*
 	return nil, nil
 }
 
-func (self StorageServiceClient) CreateStorageService(name, location string) (*StorageService, error) {
+func (self StorageServiceClient) CreateStorageService(name, location string, accountType string) (*StorageService, error) {
```
Comment (ahmetalpbalkan) [inline management/storageservice/client.go]: this method ideally should accept `StorageServiceDeployment` object, that's what I meant with my initial comment.

File: management/storageservice/entities.go [hunk 3]
```diff
@@ -34,16 +34,15 @@ type StorageServiceProperties struct {
...
-	SecondaryReadEnabled  bool
+	XMLName            xml.Name `xml:"CreateStorageServiceInput"`
+	Xmlns              string   `xml:"xmlns,attr"`
+	ServiceName        string
+	Description        string
+	Label              string
+	AffinityGroup      string `xml:",omitempty"`
+	Location           string `xml:",omitempty"`
+	ExtendedProperties ExtendedPropertyList
+	AccountType        string `xml:",omitempty"`
```
Comment (ahmetalpbalkan) [inline management/storageservice/entities.go]: move up so that it appears before ExtendedProperties.

File: management/virtualmachine/client.go [hunk 4]
```diff
@@ -131,62 +132,63 @@ func (self VirtualMachineClient) CreateAzureVMConfiguration(dnsName, instanceSiz
...
-		return nil, fmt.Errorf(errParamNotSpecified, "userName")
+		return fmt.Errorf(errParamNotSpecified, "userName")
 	}
 
 	configurationSets := ConfigurationSets{}
-	provisioningConfig, err := self.createLinuxProvisioningConfig(azureVMConfiguration.RoleName, userName, password, certPath)
+	provisioningConfig, err := self.client.createLinuxProvisioningConfig(self.DnsName, userName, password, certPath)
 	if err != nil {
-		return nil, err
+		return nil
```
Comment (ahmetalpbalkan) [inline management/virtualmachine/client.go]: return err?

File: management/virtualmachine/client.go [hunk 5]
```diff
@@ -131,62 +132,63 @@ func (self VirtualMachineClient) CreateAzureVMConfiguration(dnsName, instanceSiz
...
 	}
 
 	configurationSets.ConfigurationSet = append(configurationSets.ConfigurationSet, provisioningConfig)
 
-	networkConfig, networkErr := self.createNetworkConfig(osLinux, sshPort)
+	networkConfig, networkErr := self.client.createNetworkConfig(osLinux, sshPort)
 	if networkErr != nil {
-		return nil, err
+		return nil
```
Comment (ahmetalpbalkan) [inline management/virtualmachine/client.go]: return err?

File: management/virtualmachine/client.go [hunk 6]
```diff
@@ -214,34 +216,30 @@ func (self VirtualMachineClient) SetAzureVMExtension(azureVMConfiguration *Role,
...
-		return nil, err
+		return err
 	}
 
 	privateConfiguration := "{}"
 
-	azureVMConfiguration, err = self.SetAzureVMExtension(azureVMConfiguration, "DockerExtension", "MSOpenTech.Extensions", version, "DockerExtension", "enable", publicConfiguration, privateConfiguration)
-	return azureVMConfiguration, nil
+	err = self.SetExtension("DockerExtension", "MSOpenTech.Extensions", version, "DockerExtension", "enable", publicConfiguration, privateConfiguration)
+	return nil
```
Comment (ahmetalpbalkan) [inline management/virtualmachine/client.go]: why return nil, if there's err, we must return err

File: management/virtualmachine/entities.go [hunk 7]
```diff
@@ -11,6 +11,16 @@ type VirtualMachineClient struct {
...
 }
 
+type VMCreateConfiguration struct {
+	DnsName            string
+	InstanceSize       string
+	ImageName          string
+	Location           string
+	StorageAccountType string
+	role               *Role
+	client             VirtualMachineClient
```
Comment (ahmetalpbalkan) [inline management/virtualmachine/entities.go]: `client` definitely should not be here. having this here means a design mistake.

**Comment by ahmetalpbalkan** [comment]
My suggestion, undo the changes made to vm stuff and just focus on storage stuff in this PR. It'll be way easier for both of us to get it ready to merge.

**Comment by ahmetalpbalkan** [comment]
Closing in favor of #64 #65 #66. Thanks for bringing this up @dima-exe

**Pull request closed without merging**